services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/validation/HankePublicValidatorTest.kt

```diff
@@ -63,29 +232,86 @@ internal class HankePublicValidatorTest {
     }
 
     @Test
-    fun `No errors when rakennuttajat missing`() {
+    fun `when rakennuttajat missing should return ok`() {
         val result =
-            HankePublicValidator.validateHankeHasMandatoryFields(
+            validateHankeHasMandatoryFields(
                 completeHanke().apply { rakennuttajat = mutableListOf() }
             )
 
         assertTrue(result.isOk())
     }
 
     @Test
-    fun `No errors when toteuttajat missing`() {
+    fun `when toteuttajat missing should return ok`() {
         val result =
-            HankePublicValidator.validateHankeHasMandatoryFields(
-                completeHanke().apply { toteuttajat = mutableListOf() }
-            )
+            validateHankeHasMandatoryFields(completeHanke().apply { toteuttajat = mutableListOf() })
+
+        assertTrue(result.isOk())
+    }
+
+    @Test
+    fun `when ytunnus is present and valid should return ok`() {
+        val hanke = completeHanke()
+
+        val result = validateHankeHasMandatoryFields(hanke)
+
+        val ytunnusCount = hanke.extractYhteystiedot().mapNotNull { it.ytunnus }.count()
+        assertThat(ytunnusCount).isGreaterThanOrEqualTo(1)
+        assertTrue(result.isOk())
+    }
+
+    @Test
+    fun `when ytunnus is present and not valid should return not ok`() {
+        val hanke =
+            completeHanke().apply { rakennuttajat = rakennuttajat.modify(ytunnus = "1580375-3") }
+
+        val result = validateHankeHasMandatoryFields(hanke)
+
+        assertFalse(result.isOk())
+        assertThat(result.errorPaths()).containsExactly("rakennuttajat[0].ytunnus")
+    }
+
+    @Test
+    fun `when tyyppi yksityishenkilo or null and ytunnus is null should return ok`() {
+        val hanke =
+            completeHanke().apply {
+                omistajat = omistajat.modify(ytunnus = null, tyyppi = null)
+                rakennuttajat = rakennuttajat.modify(ytunnus = null, tyyppi = YKSITYISHENKILO)
+                toteuttajat = toteuttajat.modify(ytunnus = null, null)
+                muut = muut.modify(ytunnus = null, YKSITYISHENKILO)
+            }
+
+        val result = validateHankeHasMandatoryFields(hanke)
 
+        val ytunnusCount = hanke.extractYhteystiedot().mapNotNull { it.ytunnus }.count()
+        assertThat(ytunnusCount).isEqualTo(0)
         assertTrue(result.isOk())
     }
 
     @Test
-    fun `All errors when multiple missing fields`() {
+    fun `when tyyppi is not yksityishenkilo and ytunnus is null should not return ok`() {
+        val hanke = completeHanke().apply { toteuttajat = toteuttajat.modify(ytunnus = null) }
+
+        val result = validateHankeHasMandatoryFields(hanke)
+
+        assertFalse(result.isOk())
+        assertThat(result.errorPaths()).containsExactly("toteuttajat[0].ytunnus")
+    }
+
+    @Test
+    fun `when tyyppi is yksityishenkilo and ytunnus is not null should not return ok`() {
```

Comment:
This too might be something we want in HankeValidator. Yksityishenkilö with an y-tunnus is closer to invalid data than incomplete data.

Reply:
Added following logic to hankevalidator: 
- If ytunnus is not null, it must be valid
- If tyyppi is YKSITYISHENKILO, ytunnus must be null.